src/function.ts

```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+ * ```ts
+ * const getPerson = Function.fromFunction<string, Person>(
+ *   new aws_lambda.Function(..)
+ * );
+ *
+ * new AppsyncResolver(() => {
+ *   return getPerson("value");
+ * })
+ * ```
+ */
+export class Function<P, O> extends FunctionBase<P, O> {
+  /**
+   * Dangling promises which are processing Function handler code from the function serializer.
+   * To correctly resolve these for CDK synthesis, either use `asyncSynth()` or use `cdk synth` in the CDK cli.
+   * https://twitter.com/samgoodwin89/status/1516887131108438016?s=20&t=7GRGOQ1Bp0h_cPsJgFk3Ww
+   */
+  public static readonly promises = ((global as any)[PromisesSymbol] =
+    (global as any)[PromisesSymbol] ?? []);
+
+  /**
+   * Create a lambda function using a native typescript closure.
+   *
+   * ```ts
+   * new Function<{ val: string }, string>(this, 'myFunction', async (event) => event.val);
+   * ```
+   */
+  constructor(
+    scope: Construct,
+    id: string,
+    func: FunctionClosure<P, O>,
+    props?: Omit<aws_lambda.FunctionProps, "code" | "handler" | "runtime">
+  );
+  /**
+   * @private
+   */
+  constructor(
+    scope: Construct,
+    id: string,
+    func: NativeFunctionDecl | Err,
+    props?: Omit<aws_lambda.FunctionProps, "code" | "handler" | "runtime">
+  );
+  /**
+   * Wrap an existing lambda function with Functionless.
+   * @deprecated use `Function.fromFunction()`
+   */
+  constructor(resource: aws_lambda.IFunction);
+  /**
+   * @private
+   */
+  constructor(
+    resource: aws_lambda.IFunction | Construct,
+    id?: string,
+    func?: NativeFunctionDecl | Err | FunctionClosure<P, O>,
+    props?: Omit<aws_lambda.FunctionProps, "code" | "handler" | "runtime">
+  ) {
+    let _resource: aws_lambda.IFunction;
+    let integrations: Integration[] = [];
+    let callbackLambdaCode: CallbackLambdaCode | undefined = undefined;
+    if (func && id) {
+      if (isNativeFunctionDecl(func)) {
+        callbackLambdaCode = new CallbackLambdaCode(func.closure);
+        _resource = new aws_lambda.Function(resource, id, {
+          ...props,
+          runtime: aws_lambda.Runtime.NODEJS_14_X,
+          handler: "index.handler",
+          code: callbackLambdaCode,
+        });
+
+        integrations = func.integrations;
+      } else if (isErr(func)) {
+        throw func.error;
+      } else {
+        throw Error(
+          "Expected lambda to be passed a compiled function closure or a aws_lambda.IFunction"
+        );
+      }
+    } else {
+      _resource = resource as aws_lambda.IFunction;
+    }
+    super(_resource);
+
+    // retrieve and bootstrap all found native integrations. Will fail if the integration does not support native integration.
+    const nativeIntegrationsPrewarm = integrations
+      .map((i) => new IntegrationImpl(i))
+      .map((integration) => {
+        const integ = integration.native;
+        integ.bootstrap(this);
+        return integ;
+      })
+      .filter((i) => i.preWarm)
+      .map((i) => i.preWarm);
```

Comment:
nit: I personally think this can be simplified.

```ts
const nativeIntegrationsPrewarm = integrations.flatMap(i => {
  const integ = new IntegrationImpl(i).native;
  integ.bootstrap(this);
  return integ.preWarm ? [integ] : [];
});
```

Reply:
nice :+1: